### DISCUSSION TITLE
Missing data

### DISCUSSION THREAD

**Discussion by AlinaHovakimyan** [microsoft/gctoolkit#358] [Q&A] [answered]
Hi gctoolkit team,

I am new to the community, and faced some issues, hope you would help with them.
I am not able to get the following data from gctoolkit and was wondering what should be the next steps in such cases:
1. **VM - command line flags**:
       I see in the gctoolkit code, that it is parsing and saving this values via `Diarizer` class. But the getters seem not to be 
       implemented yet, `getCommandLine()` functions have empty bodies with only TODO's in themselves.
2. **VM version** - seems like `parseJVMVersion()` only recognizes versions less or equal to 9, so the newer versions are not supported.
3. Total **evacuation workers** count for G1 GC
     I see that the line is being parsed ( WORKER_SUMMARY parsing rule) and saved on G1GCForwardReference. But there is no 
    getter there for the `evacuationWorkersUsed` and `evacuationWorkersAvailable` variables.
3. **string& symbol table**:
     same as with worker, `stringTableProcessed`, `stringTableRemoved`, `symbolTableProcessed`, `symbolTableRemoved` are 
    being saved in the G1GCForwardReference, but i don't see getters for them.
4. And one additional question: is there a quick way to get the line number, from where the corresponding event was generated ?

I will be willing to create MR requests for some of the fixes as well, if that is the expected flow.

Thanks in advance,
Regards,
Alina

**Comment by AlinaHovakimyan** [comment]
Some more missing data:

6. Active and Total workers count for Concurrent and Full GCs
7. For G1 unified log, G1GCPauseEvent event PermOrMetaspace does not contain info about "class" and ""non-class" categories, although from the parser code i see that the info is being parsed and saved.

**Comment by Bluetopia** [comment]
For the first item, I ended up posting #336 as I noticed this behaviour too.  There's also #263 that looks for a programmatic way of determining which flags have been set for Unified logging. 

I'm not certain 2 is something that can be done with just the context of a GC log, especially past Unified Logging which occurred in Java 9.  Pre-Unified we had the log lines for JVM and command line parameters if the GC Log file didn't roll over, but I don't think I've seen similar lines in Unified logging with default GC tags. 

Can't speak to the rest, but I'd also like #5 if you end up posting a feature request for it.

**Reply by AlinaHovakimyan** [reply]
Thank you very much for the answers

**Comment by kcpeppe** [comment, marked as answer]
1) VM flags have yet to be implemented. I'm happy for any contributions on that front as it's unlikely I'll get to this for a while
2)VM version was calculated by finding features in the log file that were unique to a specific version. With unified logging the differences to identify versions basically went away so unless the log file test us the JVM version directly, then there isn't a great way to calculate this information.
3) Please add as this information is sometimes useful
4) same as 3
5) Not sure I understand the question? Are you referring to the method that produced the event?

for comment 6 and 7 please treat as 3.

**Reply by AlinaHovakimyan** [reply]
Thank you very much for the answers, 
The "_5) is there a quick way to get the line number, from where the corresponding event was generated ?_ " question was referring to the line in the gc log file, by parsing which the event was generated. Let's say I got G1PauseEvent, is there something in current implementation that would told me that the event was generated by parsing for example 10th line in the gc log-file.

**Reply by dsgrieve** [reply]
A single event may span several, possibly discontiguous, lines in the log file. You can find the line in the log using the timestamp in the event. But there is no 'getLineNumber()' API in the event itself.

**Reply by AlinaHovakimyan** [reply]
Thank you very much for the answer.
